### PR TITLE
bpo-30406: Make async and await proper keywords

### PR DESCRIPTION
**Pull request by JelleZijlstra**
https://bugs.python.org/issue30406

### CODE REVIEW

**Comment by AraHaan** [comment]
Looks, nice, However this should not break any code that uses ``async def`` if I am looking at this right?

**Review by AraHaan — APPROVED** [review]
Looks alright to me.

**Comment by JelleZijlstra** [comment]
I might need some time to fix the test failures.

**Comment by AraHaan** [comment]
Alright, But in general it looks like a nice start.

File: Lib/lib2to3/tests/test_parser.py
Comment (vstinner) [inline Lib/lib2to3/tests/test_parser.py:190]: wait... this one looks like a bug.

File: Lib/lib2to3/tests/test_parser.py
Comment (vstinner) [inline Lib/lib2to3/tests/test_parser.py:197]: ditto (also a bug)

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by 1st1** [comment]
Overall looks good. I'll make a detailed review in a few days.

**Review by vstinner — COMMENTED** [review]
https://docs.python.org/dev/reference/grammar.html should also be updated, no?

File: Lib/lib2to3/tests/test_parser.py
Comment (JelleZijlstra) [reply Lib/lib2to3/tests/test_parser.py:190]: I think lib2to3's parser in general doesn't do semantic checking of the sort that would be required to reject this. For example, it also doesn't fail on putting `nonlocal` or `yield from` outside a function, or `yield from` inside an async generator.

File: Lib/lib2to3/tests/test_parser.py
Comment (JelleZijlstra) [reply Lib/lib2to3/tests/test_parser.py:197]: Same here, this is also lib2to3.

**Comment by JelleZijlstra** [comment]
grammar.html just directly includes Grammar/Grammar, so it shouldn't require further changes.

**Comment by AraHaan** [comment]
@JelleZijlstra you mind also removing 'asyncio.async' as well as that thing might cause an SyntaxError.

**Comment by vstinner** [comment]
Thanks for the clarification :-) I let Yury review the change. He knows the
parser much better than me!

**Comment by JelleZijlstra** [comment]
@AraHaan it's currently implemented as something like `globals()['async'] = ensure_future`, so the implementation won't cause a SyntaxError (as you can tell from the test suite passing). I suppose I could remove the alias, but I also don't want to make this patch bigger than necessary, so I'll wait for Yury to review it.

**Comment by AraHaan** [comment]
Ok

**Comment by 1st1** [comment]
I'll merge this when I return from my vacation in 11 days. Thank you for keeping the pr up to date!

**Comment by JelleZijlstra** [comment]
@1st1 ping – have you had a chance to look at this PR?

**Review by 1st1 — COMMENTED** [review]
Overall looks good. Could you please rename it once again and add a news entry with blurb?

**Comment by 1st1** [comment]
s/rename/rebase

**Comment by JelleZijlstra** [comment]
Just rebased from current master and added a news entry using blurb.

**Comment by JelleZijlstra** [comment]
The AppVeyor failure is in test_idle and looks unrelated to this PR.

**Comment by JelleZijlstra** [comment]
Oh, probably related to 7c5798ebfa9b822d6cedf1fc8c034f6ea90644dc.

**Comment by vstinner** [comment]
> The AppVeyor failure is in test_idle and looks unrelated to this PR.

Rebase again and it should be fine.

File: Grammar/Grammar
Comment (1st1) [inline Grammar/Grammar]: Why did you do all these renames and added `atom_expr`? Why don't we just replace `ASYNC` with `'async'` (ditto for await)?  I'd avoid making **any** changes to the grammar besides token adjustments.

File: Lib/lib2to3/Grammar.txt
Comment (1st1) [inline Lib/lib2to3/Grammar.txt]: Adding to the point raised for changes in `Grammar`: when you change/rename/add new productions here, you might be breaking some code that uses `lib2to3` to define custom fixters. Again, I'd avoid any changes.

File: Lib/test/test_coroutines.py
Comment (1st1) [inline Lib/test/test_coroutines.py]: You can drop this comment now.

File: Misc/NEWS.d/next/Core and Builtins/2017-07-20-22-03-44.bpo-30406._kr47t.rst
Comment (1st1) [inline Misc/NEWS.d/next/Core and Builtins/2017-07-20-22-03-44.bpo-30406._kr47t.rst]: Change to:

```
Make ``async`` and ``await`` proper keywords, as specified in PEP 492.
```

**Review by 1st1 — CHANGES_REQUESTED** [review]
Playing with the code right now. One of the things you also need to do is to cleanup `forbidden_name()` function in `Python/ast.c`.

File: Grammar/Grammar
Comment (JelleZijlstra) [reply Grammar/Grammar]: I don't remember exactly (it's been a few months), but I just tried to implement your suggestions and I'm getting failures in test_parser that I don't currently understand. I'll look into it some more. The errors are all on comprehensions and are similar to

```
  File "/Users/jzijlstra-mpbt/py/cpython/Lib/test/test_parser.py", line 24, in roundtrip
    self.fail("could not roundtrip %r: %s" % (s, why))
AssertionError: could not roundtrip '{x for x in seq}': Illegal terminal: expected NAME, got 330.
```

**Comment by 1st1** [comment]
Ping? I plan to work on this myself if this isn't resolved soon.

**Comment by JelleZijlstra** [comment]
I was unable to resolve the issue I mentioned in https://github.com/python/cpython/pull/1669#discussion_r128901454. I can push out the other fixes though and take another look within the next few days to see if I can get it working.

**Comment by 1st1** [comment]
> I was unable to resolve the issue I mentioned in #1669 (comment). [..]

Interesting. I'll take a look at that too.

**Comment by 1st1** [comment]
So here's my attempt at making this work without extensive grammar changes: https://github.com/1st1/cpython/commit/0eee314f87607fb7cf5178cd0998a9825070af18

If you are interested in continuing to work on this PR, I suggest you to incorporate my approach.

**Comment by JelleZijlstra** [comment]
Thanks! I'll spend some time on it tonight (California time).

**Comment by JelleZijlstra** [comment]
My attempt was very similar to yours, but your code also fails test_parser (at least when I compile it locally). I'll push what I have for now.

**Comment by 1st1** [comment]
> My attempt was very similar to yours, but your code also fails test_parser (at least when I compile it locally).

I didn't run the tests, just checked if it was possible to fix the grammar in the minimal way. Some tests should probably be adjusted.

Please also make sure that lib2to3/grammar is in sync with grammar/grammar.

**Comment by JelleZijlstra** [comment]
I fixed lib2to3 and adjusted all tests that were obvious.

Unfortunately the `test_parser` problem doesn't look like it's just a matter of adjusting the test. The failures look like this:
```
======================================================================
FAIL: test_set_comprehensions (test.test_parser.RoundtripLegalSyntaxTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_parser.py", line 22, in roundtrip
    st2 = parser.sequence2st(t)
parser.ParserError: Illegal terminal: expected NAME.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_parser.py", line 400, in test_set_comprehensions
    self.check_expr('{x for x in seq}')
  File "/home/travis/build/python/cpython/Lib/test/test_parser.py", line 30, in check_expr
    self.roundtrip(parser.expr, s)
  File "/home/travis/build/python/cpython/Lib/test/test_parser.py", line 24, in roundtrip
    self.fail("could not roundtrip %r: %s" % (s, why))
AssertionError: could not roundtrip '{x for x in seq}': Illegal terminal: expected NAME.
```
I couldn't figure out why this doesn't work when all other tests pass. Something must be wrong related to the code for async comprehensions but I'm not sure what.

**Comment by 1st1** [comment]
> Unfortunately the test_parser problem doesn't look like it's just a matter of adjusting the test. The failures look like this:

OK, I think I figured this out.  The problem is that when we have `comp_for` defined as follows:

```
comp_for: ['async'] 'for' exprlist 'in' or_test [comp_iter]
```

we have two states in the grammar for `comp_for`:

```
state1: NAME NAME exprlist NAME or_test ...
state2: NAME exprlist NAME or_test ...
```

and the code in `parsermodule.c` (`validate_node()` function) doesn't know how to pick the correct state from all possible states in the generated grammar.  The code in `ast.c` doesn't have this problem as we hardcode states traversal logic.  Theoretically we could fix `validate_node()`, but I'm afraid it would complicate things too much, so let's just add another non-term to the grammar.

I propose to split `comp_for` into two non-terms: `sync_comp_for` and `comp_for`:

```
sync_comp_for: 'for' exprlist 'in' or_test [comp_iter]
comp_for: ['async'] sync_comp_for
```

This seems to be less of a change compared to the first approach in this PR (adding `comp_for_or_async` and `comp_async_for`).

I tried it here https://github.com/1st1/cpython/commit/3457053825d6f96662b816514be11e126cc1014b and it seems to work for all `test_parser` tests.

**Comment by JelleZijlstra** [comment]
Thanks for looking into this! The suggested change makes sense. I'll implement that in the next few days and make sure all tests pass (might not have time today).

**Comment by 1st1** [comment]
> I'll implement that in the next few days and make sure all tests pass (might not have time today).

No rush, I won't be able to further review till next Monday anyways.

**Comment by JelleZijlstra** [comment]
Just pushed an update that passes all tests locally. I'll check in in a while to make sure Travis passes.

File: Doc/library/token.rst
Comment (1st1) [inline Doc/library/token.rst]: Drop "again".

File: Python/ast.c
Comment (1st1) [inline Python/ast.c:1853]: I think we should add another macro:

```C
#define REQ_NAME(node, name) \
    REQ(node, NAME); \
    assert(strcmp(STR(node), name) == 0);
```

so that 

```C
REQ(CHILD(n, 0), NAME);
assert(strcmp(STR(CHILD(n, 0)), "async") == 0);
```

becomes

```C
REQ_NAME(CHILD(n, 0), "async");
```

**Comment by 1st1** [comment]
Looks good. @haypo I'll merge this unless you have more comments.

File: Python/ast.c
Comment (JelleZijlstra) [reply Python/ast.c:1853]: Should that go into Include/node.h? I can add it but I'm not sure it deserves to be in the public API, and it seems out of scope for this PR.

File: Doc/library/token.rst
Comment (JelleZijlstra) [reply Doc/library/token.rst]: Done.

File: Python/ast.c
Comment (1st1) [reply Python/ast.c:1853]: IIRC "node.h" isn't part of the public API at all. But it's fine, let's skip this bit.

Comment (JelleZijlstra) [reply Python/ast.c:1853]: OK, I can make another PR after this one is merged.

File: Doc/library/token.rst
Comment (vstinner) [inline Doc/library/token.rst:137]: Why not saying "keywords". These names are added to keywords.py and it's the PR title, no?

**Review by vstinner — APPROVED** [review]
LGTM with a minor question on the doc.

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
@JelleZijlstra Could you please rebase against current master?  I resolved one conflict in lib2to3/Grammar, but there're few other things that broke CI.

**Comment by JelleZijlstra** [comment]
Doing that now. Are you sure there is another conflict though? GitHub claims there isn't.

**Comment by JelleZijlstra** [comment]
Just rebased and adjusted the lib2to3 grammar changes to make local tests pass.

**Comment by ambv** [comment]
@JelleZijlstra This change is incorrect in lib2to3. That library's grammar is explicitly kept in a way that is able to parse all code valid for Python 2 - Python 3.7. With your change we lost the ability to read valid Python 2 - Python 3.6 code that uses `async` and `await` as names.

**Comment by 1st1** [comment]
Hm, I'm not sure what options do we have here. Any suggestions?  About the only one that I have is to undo the change for lib2to3 entirely (but keeping the rest of the changes).

**Comment by ambv** [comment]
Yes, this is what I'm inclined to do. Revert the lib2to3 part.

**Comment by JelleZijlstra** [comment]
If I recall correctly I made changes to lib2to3 because some test was failing otherwise, so it might not be as simple as just reverting parts of this PR. I can spend some time looking into this later today.

**Comment by JelleZijlstra** [comment]
OK, I made that up—tests pass cleanly if I undo my lib2to3 changes. @ambv @1st1 I submitted #6143 to fix the regression.

**Pull request merged**
